### ISSUE TITLE
Inline temporary variable changes the behavior

### ISSUE THREAD

**Issue by alrz** [dotnet/roslyn#16230]
**Version Used**: VS2015 Update 3 - 14.0.25420.01

**Steps to Reproduce**:

```cs
int F()
{
  var [||]result = 1;
  if(G()) return result;
  result = 2;
  return result;
}
```

**Expected Behavior**:
```cs
int F()
{
  if(G()) return 1;
  return 2;
}
```
**Actual Behavior**:
```cs
int F()
{
  var result = 1;
  if(G()) return 1;
  result = 2;
  return 1;
}
```

**Comment by Pilchie** [comment]
Yes, but you also get big red conflict indicators in the preview...

@DustinCampbell

**Comment by alrz** [comment]
@Pilchie Shouldn't be at least missing? i.e. not suggesting the fix.

**Comment by alrz** [comment]
I didn't check the pewview btw. In my case there was a lot of other code in between and it could be easily overlooked. That's not safe.

**Comment by Pilchie** [comment]
In general, our philosophy is to make refactorings available, and assume that the user knows what they are doing, and instead just try to _inform_ about the consequences using the preview.

**Comment by alrz** [comment]
That seems to be in disagreement with "behavior must not be changed after the fix is applied" policy which is reported otherwise on this repo a couple of times and get fixed.

**Comment by Pilchie** [comment]
We don't *want* to change behavior, and we certainly don't want analyzers suggesting that you *should* change behavior, but in cases where there is no way to do both what the refactoring is designed to do and to not change behavior, we will change behavior with a conflict marker.

**Comment by DustinCampbell** [comment]
To be fair, the refactoring *should* detect that both assignments to the variable are immediately returned and simply remove the local. That's just a better refactoring. :smile:

IMO, we should keep this bug.

**Comment by DustinCampbell** [comment]
> A similar case could be: "the field X is always assigned before being used and can be converted to a local"

I'm not sure I fully understand what you meant by that or why it's similar to this. Could you provide an example?

**Comment by alrz** [comment]
For example:
```cs
class X
{
    private object _o;
    void F()
    {
        _o = new object();
        // use _o
    }
    void G()
    {
        _o = new object();
        // use _o
    }
}
```
That field doesn't need to be a field. It is a similar case because again you are assigning it immediately before use. Depending on the code that comes after the assignment, it again could be simplified further to be inlined.

**Comment by DustinCampbell** [comment]
I understand now. I would see that as a different refactoring. Determining whether a field is needed or not is much more complex.

**Comment by alrz** [comment]
About this issue, I agree it does require a smarter analysis to figure out the case mentioned in OP but I think it would make sense to keep it in the backlog at least.. .
